### PR TITLE
Speed up CSpace1_R proofs

### PR DESCRIPTION
**Pull request by lsf37**
Not finished yet, but maybe useful to look at already before we apply this to the other architectures and MCS.

There is also more work to be done still in the second CSpace theory (`CSpace_R`), but I hope that the same ideas will apply.

I haven't measured the actual time gain, but it feels interactively reasonably fast now, whereas before it was definitely in "get a cup of tea" territory.

(this is accidentally on top of #451, which hopefully is merged soon and will disappear here).

### CODE REVIEW

**Comment by lsf37** [comment]
Ok, have added a pass through CSpace_R as well, and it doesn't seem to be taking ages for any particular lemma any more.

**Comment by lsf37** [comment]
Timings:

```
Refine 2021-1 this PR: 0:23:56
Refine 2021-1 before:  0:56:53
Refine 2021:           0:18:32
```

So we're back to at least the same order of magnitude for this one. `CSpace*_R` should now be faster than in the 2021 version, but there are probably many places in other proofs that in accumulation lead to 5 min longer real time.

We should also check the other sessions, I haven't done that yet.

**Comment by lsf37** [comment]
Data for the other sessions and architectures (ignore the Refine numbers, they are without this PR). We should merge this PR first before we look at any of the others.

---
**ARM**
```
tests-xml-correct       +0:00 (  +0.0%) real,  +0:00 (  +0.0%) cpu,  +0.00GB (  +0.0%)
haskell-translator      +0:00 (  +0.0%) real,  -1:59 ( -14.3%) cpu,  +0.00GB (  +0.0%)
HaskellKernel           +0:15 (  +3.7%) real,  -2:20 (  -4.7%) cpu,  -0.03GB (  -3.7%)
isabelle                +0:15 (  +3.4%) real,  +0:49 (  +3.0%) cpu,  -0.13GB (  -1.5%)
Docs                    +0:01 (  +7.7%) real,  +0:03 (  +9.7%) cpu,  -0.12GB (  -8.6%)
Lib                     +0:11 ( +22.4%) real,  +0:26 (  +8.2%) cpu,  +1.80GB ( +57.3%)
CamkesGlueSpec          +0:00 (  +0.0%) real,  +0:01 (  +1.2%) cpu,  -0.02GB (  -0.9%)
Sep_Algebra             +0:07 ( +17.5%) real,  +0:25 ( +10.0%) cpu,  -2.96GB ( -47.1%)
CParser                 -1:59 (  -0.7%) real,  -1:46 (  -2.5%) cpu,  +0.25GB (  +5.5%)
Concurrency             +0:05 (  +5.6%) real,  -1:50 (  -2.1%) cpu,  +2.69GB ( +80.3%)
EVTutorial              +0:12 ( +15.2%) real,  +0:32 (  +8.4%) cpu,  -0.58GB ( -15.4%)
TakeGrant               +0:03 ( +11.5%) real,  +0:10 ( +14.7%) cpu,  +0.16GB (  +8.3%)
SepTactics              +0:06 (  +9.4%) real,  +0:14 (  +5.0%) cpu,  +2.16GB ( +35.6%)
ASpec                   +0:08 (  +4.2%) real,  +0:06 (  +1.4%) cpu,  +0.75GB ( +21.5%)
ExecSpec                -1:56 (  -1.7%) real,  -2:47 ( -12.7%) cpu,  +2.69GB ( +57.7%)
DSpec                   -1:57 (  -2.4%) real,  -1:51 (  -2.8%) cpu,  +1.35GB ( +27.3%)
SepDSpec                +0:08 ( +15.1%) real,  +0:22 (  +8.9%) cpu,  +0.07GB (  +1.6%)
DSpecProofs             +0:05 (  +6.5%) real,  +0:24 (  +6.0%) cpu,  -0.80GB ( -27.3%)
ASepSpec                -1:57 ( -12.5%) real,  -1:56 (  -7.5%) cpu,  +0.03GB (  +1.8%)
SysInit                 +0:30 ( +19.4%) real,  +2:06 ( +17.6%) cpu,  +0.06GB (  +1.6%)
SysInitExamples         +0:30 ( +13.6%) real,  +1:28 ( +11.8%) cpu,  +0.24GB (  +8.9%)
CLib                    -1:59 (  -2.3%) real,  -1:56 (  -1.4%) cpu,  +1.16GB ( +36.7%)
c-kernel                -1:58 (  -7.4%) real,  -1:53 ( -15.2%) cpu,  +0.16GB ( +12.1%)
AInvs                   +6:06 ( +36.9%) real, +13:50 ( +18.6%) cpu,  +0.53GB (  +3.6%)
BaseRefine              +1:07 ( +32.1%) real,  +3:01 ( +46.2%) cpu,  -1.14GB ( -19.4%)
CKernel                 -3:52 ( -10.6%) real,  -6:44 (  -7.5%) cpu,  +2.47GB ( +13.5%)
DBaseRefine             -1:59 (  -0.6%) real,  +0:04 (  +1.3%) cpu,  +0.09GB (  +1.8%)
Refine                 +33:18 (+141.2%) real, +63:08 ( +48.4%) cpu,  -2.76GB ( -12.6%)
DRefine                 +1:00 ( +12.3%) real,  +4:35 ( +10.2%) cpu,  +0.13GB (  +1.7%)
RefineOrphanage         +0:08 (  +8.7%) real,  +0:08 (  +3.6%) cpu,  +0.17GB (  +4.5%)
DPolicy                 -1:49 (  -6.5%) real,  -2:54 ( -15.3%) cpu,  +0.47GB ( +11.4%)
Bisim                   +0:04 (  +9.3%) real,  +0:04 (  +3.8%) cpu,  +0.06GB (  +2.0%)
Access                  +1:09 ( +16.4%) real,  +3:36 ( +12.6%) cpu,  +2.48GB ( +24.6%)
LibTest                 +0:09 (  +6.0%) real,  +0:15 (  +2.2%) cpu,  +0.50GB (  +4.8%)
CamkesAdlSpec           +0:01 (  +1.0%) real,  +0:04 (  +2.4%) cpu,  -0.06GB (  -1.6%)
CamkesCdlRefine         -1:51 (  -6.0%) real,  -1:35 (  -8.7%) cpu,  +0.36GB ( +10.1%)
CSpec                   -2:36 ( -20.9%) real,  -8:00 ( -24.2%) cpu,  +0.78GB (  +5.3%)
InfoFlow                -1:59 (  -0.1%) real,  -2:51 (  -2.1%) cpu,  +0.13GB (  +0.6%)
CBaseRefine             -1:22 (  -2.8%) real,  -8:17 ( -13.5%) cpu,  +0.76GB (  +3.5%)
SimplExportAndRefine    +0:21 (  +0.8%) real, +11:04 (  +4.6%) cpu, +10.42GB ( +92.5%)
CParserTest             -1:32 (  -6.3%) real,  -1:36 (  -1.1%) cpu,  -5.22GB ( -29.2%)
CParserTools            -1:50 (  -6.8%) real,  -1:50 (  -7.1%) cpu,  +0.00GB (  +0.0%)
AutoCorres              +0:02 (  +3.0%) real,  -1:50 (  -2.8%) cpu,  +0.94GB ( +26.5%)
CamkesGlueProofs        +2:19 ( +22.2%) real,  +8:00 ( +23.0%) cpu,  -1.29GB ( -18.7%)
AutoCorresDoc           +0:00 (  +0.0%) real,  +0:00 (  +0.0%) cpu,  -0.05GB (  -2.4%)
AutoCorresTest          +1:32 ( +36.2%) real,  +4:55 ( +18.0%) cpu,  +2.57GB ( +17.4%)
AsmRefine               -1:51 (  -5.7%) real,  -1:11 ( -12.8%) cpu,  +1.01GB ( +38.4%)
CRefine                 +7:56 ( +13.0%) real, +44:48 ( +14.6%) cpu,  +0.52GB (  +2.3%)
InfoFlowCBase           +0:06 (  +1.3%) real,  -1:29 (  -3.9%) cpu,  +1.75GB ( +16.7%)
InfoFlowC               +0:46 (  +5.7%) real,  +4:38 ( +12.3%) cpu,  -5.53GB ( -25.7%)
```
---
**ARM_HYP**
```
tests-xml-correct       +0:00 (  +0.0%) real,  +0:00 (  +0.0%) cpu,  +0.00GB (  +0.0%)
haskell-translator      +0:01 (  +7.7%) real,  -1:59 ( -14.3%) cpu,  +0.00GB (  +0.0%)
HaskellKernel           +0:09 (  +2.2%) real,  -4:58 (  -8.2%) cpu,  +0.04GB (  +5.3%)
isabelle                +0:25 (  +5.7%) real,  +0:17 (  +1.1%) cpu,  +1.09GB ( +15.2%)
Docs                    +0:01 (  +7.7%) real,  +0:05 ( +15.6%) cpu,  -0.10GB (  -7.1%)
Lib                     +0:15 ( +31.2%) real,  +0:40 ( +12.8%) cpu,  -2.28GB ( -42.2%)
CamkesGlueSpec          +0:00 (  +0.0%) real,  -1:59 (  -1.2%) cpu,  +0.04GB (  +2.0%)
Sep_Algebra             +0:09 ( +23.1%) real,  +0:35 ( +14.8%) cpu,  -0.11GB (  -3.5%)
CParser                 +0:05 (  +3.4%) real,  +0:08 (  +1.4%) cpu,  +0.35GB (  +7.4%)
Concurrency             +0:05 (  +5.4%) real,  -1:53 (  -1.4%) cpu,  +1.88GB ( +44.8%)
EVTutorial              +0:24 ( +30.4%) real,  +1:15 ( +19.6%) cpu,  -0.28GB (  -8.4%)
TakeGrant               +0:03 ( +12.0%) real,  +0:10 ( +15.6%) cpu,  +0.16GB (  +8.3%)
SepTactics              +0:09 ( +14.3%) real,  +0:19 (  +6.7%) cpu,  +2.67GB ( +83.2%)
ASpec                   +0:15 (  +6.0%) real,  +0:08 (  +1.4%) cpu,  +1.59GB ( +36.5%)
ExecSpec                +0:31 ( +10.0%) real,  +0:41 (  +5.5%) cpu,  +2.84GB ( +58.8%)
DSpec                   +0:02 (  +1.6%) real,  +0:16 (  +4.4%) cpu,  +7.36GB (+142.9%)
SepDSpec                +0:10 ( +20.8%) real,  +0:38 ( +16.7%) cpu,  +1.13GB ( +35.2%)
DSpecProofs             +0:04 (  +5.1%) real,  +0:16 (  +4.0%) cpu,  +0.38GB ( +16.0%)
ASepSpec                +0:01 (  +4.0%) real,  -1:57 (  -5.2%) cpu,  +0.10GB (  +5.6%)
SysInit                 +0:42 ( +26.8%) real,  +2:41 ( +22.1%) cpu,  -2.56GB ( -44.6%)
SysInitExamples         +0:34 ( +15.5%) real,  +1:45 ( +14.8%) cpu,  +0.22GB (  +7.8%)
CLib                    +0:02 (  +4.4%) real,  +0:07 (  +2.5%) cpu,  -0.59GB ( -13.6%)
c-kernel                +0:06 ( +27.3%) real,  +0:10 ( +25.0%) cpu,  -0.18GB ( -10.7%)
AInvs                   +8:15 ( +35.8%) real, +23:17 ( +25.4%) cpu,  -3.98GB ( -18.8%)
BaseRefine              +0:57 ( +18.6%) real,  +1:10 ( +11.0%) cpu,  -2.54GB ( -30.1%)
CKernel                 +0:15 (  +1.0%) real,  +2:23 (  +2.5%) cpu,  +6.77GB ( +48.8%)
DBaseRefine             +0:02 (  +1.2%) real,  +0:03 (  +1.0%) cpu,  +0.83GB ( +14.6%)
CSpec                   -2:37 ( -11.6%) real,  -9:37 ( -13.9%) cpu,  +1.72GB (  +9.8%)
Refine                 +26:41 ( +77.9%) real, +47:37 ( +26.8%) cpu,  +4.73GB ( +26.8%)
LibTest                 +0:33 ( +30.0%) real,  +1:13 ( +11.6%) cpu,  -4.63GB ( -26.1%)
CParserTest             +3:34 ( +56.3%) real, +12:58 ( +42.1%) cpu,  -5.91GB ( -28.4%)
CParserTools            +0:05 (  +3.6%) real,  +0:04 (  +3.0%) cpu,  -0.01GB (  -0.2%)
AutoCorres              +0:16 ( +28.6%) real,  +1:00 ( +19.7%) cpu,  +0.18GB (  +5.0%)
CamkesGlueProofs        +3:02 ( +31.7%) real, +10:11 ( +32.7%) cpu,  -2.57GB ( -34.7%)
AutoCorresDoc           +0:02 ( +13.3%) real,  +0:06 ( +14.3%) cpu,  -0.02GB (  -1.0%)
AutoCorresTest          +0:17 (  +6.0%) real,  +2:37 (  +9.3%) cpu,  +3.15GB ( +29.7%)
CBaseRefine             +5:28 ( +15.7%) real,  +9:42 ( +10.6%) cpu,  +0.01GB (  +0.0%)
AsmRefine               +0:09 (  +6.2%) real,  -1:55 (  -1.4%) cpu,  -1.97GB ( -31.9%)
CRefine                +16:27 ( +21.3%) real, +92:48 ( +25.1%) cpu,  +0.77GB (  +3.4%)
```
---
**RISCV64**
```
tests-xml-correct       +0:00 (  +0.0%) real,  +0:00 (  +0.0%) cpu,  +0.00GB (  +0.0%)
haskell-translator      +0:00 (  +0.0%) real,  +0:01 ( +16.7%) cpu,  +0.00GB (  +0.0%)
HaskellKernel           +0:07 (  +1.6%) real,  +0:33 (  +1.4%) cpu,  +0.00GB (  +0.0%)
isabelle                +0:19 (  +4.1%) real,  +1:04 (  +3.9%) cpu,  -0.13GB (  -1.5%)
Docs                    +0:02 ( +14.3%) real,  +0:07 ( +21.2%) cpu,  -0.02GB (  -1.4%)
Lib                     +0:15 ( +31.2%) real,  +0:58 ( +18.4%) cpu,  +0.64GB ( +20.7%)
CamkesGlueSpec          +0:05 ( +15.6%) real,  +0:11 ( +13.4%) cpu,  +0.08GB (  +3.9%)
Sep_Algebra             +0:12 ( +31.6%) real,  +0:52 ( +22.1%) cpu,  -0.74GB ( -14.1%)
CParser                 +0:15 ( +10.1%) real,  +0:53 (  +9.7%) cpu,  +1.26GB ( +27.0%)
Concurrency             +0:08 (  +8.7%) real,  +0:08 (  +1.7%) cpu,  +1.41GB ( +35.7%)
EVTutorial              +0:13 ( +16.5%) real,  +0:37 (  +9.7%) cpu,  +1.14GB ( +36.0%)
TakeGrant               +0:04 ( +14.8%) real,  +0:11 ( +15.9%) cpu,  +0.04GB (  +2.0%)
SepTactics              +0:10 ( +15.2%) real,  +0:35 ( +12.6%) cpu,  -1.02GB ( -21.2%)
ASpec                   +0:42 ( +22.3%) real,  +1:59 ( +28.9%) cpu,  -0.43GB (  -9.0%)
ExecSpec                +0:22 (  +9.0%) real,  +0:27 (  +4.5%) cpu,  +0.83GB ( +21.6%)
ASepSpec                +0:03 ( +14.3%) real,  +0:09 ( +19.1%) cpu,  +0.12GB (  +7.4%)
CLib                    +0:01 (  +2.0%) real,  +0:23 (  +7.8%) cpu,  -0.13GB (  -3.1%)
c-kernel                +0:05 ( +20.0%) real,  +0:11 ( +27.5%) cpu,  +0.05GB (  +3.3%)
AInvs                   +2:38 ( +23.4%) real, +13:41 ( +22.6%) cpu,  -5.36GB ( -25.0%)
BaseRefine              +0:36 ( +16.1%) real,  +1:29 ( +20.9%) cpu,  -0.56GB ( -10.1%)
CKernel                 -1:44 (  -2.0%) real,  +0:05 (  +0.2%) cpu,  -2.69GB ( -15.6%)
Access                  +3:42 ( +39.4%) real,  +7:51 ( +18.0%) cpu,  -1.89GB ( -20.5%)
CamkesAdlSpec           -1:52 (  -7.7%) real,  -1:48 (  -6.9%) cpu,  -0.35GB (  -8.1%)
Refine                 +25:42 ( +96.7%) real, +49:46 ( +35.2%) cpu,  -0.33GB (  -1.5%)
RefineOrphanage         +0:08 (  +8.6%) real,  +0:09 (  +4.1%) cpu,  +0.20GB (  +5.4%)
Bisim                   +0:03 (  +7.0%) real,  +0:02 (  +1.9%) cpu,  +0.09GB (  +2.9%)
LibTest                 +0:18 ( +15.9%) real,  +0:33 (  +5.2%) cpu,  -1.69GB ( -16.1%)
CSpec                   -1:19 ( -14.9%) real,  -4:30 ( -15.7%) cpu,  -1.77GB ( -22.5%)
InfoFlow                +0:36 (  +2.2%) real,  -2:32 (  -1.6%) cpu, +10.58GB ( +98.1%)
CBaseRefine             +3:03 ( +15.1%) real,  +5:52 ( +13.1%) cpu,  +0.47GB (  +2.6%)
SimplExportAndRefine    +1:41 (  +4.9%) real, +15:45 (  +8.1%) cpu,  +9.34GB ( +77.7%)
CParserTest             -2:57 ( -13.3%) real,  -3:55 (  -5.5%) cpu,  +0.63GB (  +3.2%)
CParserTools            -1:50 (  -6.6%) real,  -1:51 (  -6.2%) cpu,  -0.33GB (  -6.4%)
AutoCorres              +0:04 (  +6.1%) real,  +0:36 ( +10.8%) cpu,  -0.48GB ( -13.4%)
AutoCorresDoc           +0:01 (  +6.2%) real,  +0:03 (  +6.7%) cpu,  -0.02GB (  -1.0%)
AutoCorresTest          -1:39 (  -6.6%) real,  -1:02 (  -3.1%) cpu,  +1.69GB ( +14.6%)
CRefine                +10:01 ( +20.4%) real, +34:17 ( +13.6%) cpu,  +0.17GB (  +0.7%)
InfoFlowCBase           +0:47 ( +12.5%) real,  +2:02 ( +19.4%) cpu,  -0.05GB (  -0.5%)
InfoFlowC               +5:54 ( +31.7%) real, +12:51 ( +20.4%) cpu,  -4.33GB ( -28.0%)
```
---
**X64**
```
tests-xml-correct       +0:00 (  +0.0%) real,  +0:00 (  +0.0%) cpu,  +0.01GB (  +0.0%)
haskell-translator      +0:02 ( +16.7%) real,  +0:01 ( +16.7%) cpu,  +0.00GB (  +0.0%)
HaskellKernel           -1:49 (  -2.6%) real,  -3:11 (  -7.7%) cpu,  +0.00GB (  +0.0%)
isabelle                +0:02 (  +0.4%) real,  -1:49 (  -0.7%) cpu,  +0.67GB (  +7.5%)
Docs                    +0:02 ( +16.7%) real,  +0:04 ( +13.8%) cpu,  +0.17GB ( +15.5%)
Lib                     +0:13 ( +28.3%) real,  +0:39 ( +13.0%) cpu,  -0.30GB (  -7.0%)
CamkesGlueSpec          +0:02 (  +6.5%) real,  +0:07 (  +9.5%) cpu,  -0.70GB ( -26.1%)
Sep_Algebra             +0:12 ( +33.3%) real,  +0:50 ( +22.1%) cpu,  -1.42GB ( -32.9%)
CParser                 +0:08 (  +5.6%) real,  +0:30 (  +5.7%) cpu,  +0.20GB (  +4.7%)
Concurrency             +0:12 ( +14.3%) real,  +0:23 (  +5.1%) cpu,  -1.76GB ( -32.6%)
EVTutorial              +0:16 ( +21.9%) real,  +0:57 ( +16.3%) cpu,  +0.95GB ( +23.6%)
TakeGrant               +0:06 ( +25.0%) real,  +0:15 ( +23.8%) cpu,  -0.06GB (  -3.0%)
SepTactics              +0:13 ( +22.0%) real,  +0:40 ( +15.3%) cpu,  +0.52GB ( +17.5%)
ASpec                   +0:24 (  +9.9%) real,  +0:31 (  +5.6%) cpu,  -1.39GB ( -17.4%)
ExecSpec                +0:35 ( +11.7%) real,  +0:45 (  +6.4%) cpu,  +0.94GB ( +13.0%)
ASepSpec                +0:00 (  +0.0%) real,  +0:01 (  +2.0%) cpu,  +0.14GB (  +8.0%)
CLib                    +0:06 ( +15.4%) real,  +0:23 (  +8.7%) cpu,  -0.46GB ( -11.4%)
c-kernel                +0:00 (  +0.0%) real,  +0:01 (  +2.2%) cpu,  -0.20GB ( -10.6%)
AInvs                   +3:30 ( +23.3%) real, +18:32 ( +24.2%) cpu,  -0.28GB (  -1.3%)
BaseRefine              +0:42 ( +13.1%) real,  +0:46 (  +7.1%) cpu,  +1.86GB ( +30.7%)
CKernel                 -5:29 ( -10.2%) real, -27:29 ( -17.7%) cpu,  +0.15GB (  +0.7%)
Bisim                   +0:16 ( +32.0%) real,  +0:20 ( +17.4%) cpu,  +0.02GB (  +0.6%)
CSpec                   +0:21 (  +3.1%) real,  -3:44 (  -3.8%) cpu,  +1.50GB (  +8.7%)
Refine                 +13:18 ( +27.8%) real, +44:40 ( +22.9%) cpu,  +0.51GB (  +2.3%)
LibTest                 +0:56 ( +51.9%) real,  +2:09 ( +21.6%) cpu,  -4.61GB ( -30.4%)
CParserTest             +2:53 ( +48.3%) real, +10:55 ( +36.5%) cpu,  -3.17GB ( -15.5%)
CParserTools            +0:07 (  +5.3%) real,  +0:06 (  +4.7%) cpu,  +0.02GB (  +0.4%)
AutoCorres              +0:06 ( +10.5%) real,  +0:39 ( +12.7%) cpu,  +1.87GB ( +56.3%)
AutoCorresDoc           +0:01 (  +6.2%) real,  +0:06 ( +15.4%) cpu,  +0.05GB (  +2.5%)
AutoCorresTest          +1:37 ( +34.9%) real,  +5:32 ( +19.7%) cpu,  +0.51GB (  +4.7%)
CBaseRefine            +10:24 ( +38.3%) real, +33:01 ( +49.4%) cpu,  +0.95GB (  +4.3%)
CRefine                +25:16 ( +33.7%) real, +87:35 ( +24.7%) cpu,  +0.47GB (  +2.0%)
```

File: proof/refine/X64/CSpace1_R.thy [hunk 1]
```diff
@@ -2905,32 +2892,25 @@ lemma is_derived_eq:
...
-  apply (simp add:vs_cap_ref_def)
+  subgoal by (cases c, auto simp: isCap_simps cap_master_cap_def capMasterCap_def vsCapRef_def
+                                   vs_cap_ref_def)[1]
   apply (case_tac "isIRQControlCap d'")
    apply (frule(1) master_cap_relation)
-   apply (clarsimp simp: isCap_simps cap_master_cap_def
-                         is_zombie_def is_reply_cap_def is_master_reply_cap_def
-                  split: cap_relation_split_asm arch_cap.split_asm)[1]
+  subgoal by (clarsimp simp: isCap_simps cap_master_cap_def
+                        is_zombie_def is_reply_cap_def is_master_reply_cap_def
```
Comment (michaelmcinerney) [inline proof/refine/X64/CSpace1_R.thy:2900]: Nitpick: indentation

File: proof/refine/X64/CSpace1_R.thy [hunk 2]
```diff
@@ -2905,32 +2892,25 @@ lemma is_derived_eq:
   apply (clarsimp simp: is_derived_def is_derived'_def badge_derived'_def)
   apply (rule conjI)
    apply (clarsimp simp: is_cap_simps isCap_simps)
-   apply (cases c, auto simp: isCap_simps cap_master_cap_def capMasterCap_def)[1]
-  apply (simp add:vsCapRef_def)
-  apply (simp add:vs_cap_ref_def)
+  subgoal by (cases c, auto simp: isCap_simps cap_master_cap_def capMasterCap_def vsCapRef_def
+                                   vs_cap_ref_def)[1]
```
Comment (michaelmcinerney) [inline proof/refine/X64/CSpace1_R.thy:2896]: Nitpick: indentation

**Review by michaelmcinerney — APPROVED** [review]
Looks great, thanks Gerwin

File: proof/refine/ARM/CSpace1_R.thy [hunk 3]
```diff
@@ -6603,24 +6613,49 @@ lemma cteSwap_corres:
...
+  apply (thin_tac "ksArchState t = p" for t p)+
+  apply (thin_tac "gsCNodes t = p" for t p)+
+  apply (thin_tac "ksWorkUnitsCompleted t = p" for t p)+
+  apply (thin_tac "ksInterruptState t = p" for t p)+
+  apply (thin_tac "ksIdleThread t = p" for t p)+
+  apply (thin_tac "gsUserPages t = p" for t p)+
+  apply (thin_tac "pspace_relation s s'" for s s')+
+  apply (thin_tac "ekheap_relation e p" for e p)+
+  apply (thin_tac "interrupt_state_relation n s s'" for n s s')+
+  apply (thin_tac "(s,s') \<in> arch_state_relation" for s s')+
```
Comment (Xaphiosis) [inline proof/refine/ARM/CSpace1_R.thy:6644]: While the performance is good, this is a bit sad to have to have in here.
I guess if the goal gets too big, there is no other way :/

**Review by Xaphiosis — APPROVED** [review]

**Comment by Xaphiosis** [comment]
Don't forget to squash (you won't get a warning here). Incidentally, we should add "sqash" to whatever warning system has "squash" and "wip", because I've seen this typo before.

**Comment by lsf37** [comment]
> Don't forget to squash (you won't get a warning here). 

:+1:

> Incidentally, we should add "sqash" to whatever warning system has "squash" and "wip", because I've seen this typo before.

Good idea, I'll have a look at that.

File: proof/refine/ARM/CSpace1_R.thy (hunk 3 above)
Comment (lsf37) [reply proof/refine/ARM/CSpace1_R.thy:6644]: Another way would be to use Isar and state the bit we want to prove explicitly, but those are also very big and quite ugly, so I think we're stuck with this method for now.

**Pull request merged**